### ISSUE TITLE
CONFIG_ADC_CONTINUOUS_ISR_IRAM_SAFE is only partially implemented in ADC continuous driver 

### ISSUE THREAD

**Issue by ivmarkov** [espressif/esp-idf#12117]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

master

### Operating System used.

Linux

### How did you build your project?

Other (please specify in More Information)

### If you are using Windows, please specify command line type.

None

### Development Kit.

Any

### Power Supply used.

USB

### What is the expected behavior?

When `CONFIG_ADC_CONTINUOUS_ISR_IRAM_SAFE` is set to NO ('n') - which is the default anyway, I would expect that it would be OK to put my callback handlers in flash.

However, the ADC continuous driver code is instantiating the interrupt handler unconditionally **and always** with `ESP_INTR_FLAG_IRAM` - [here](https://github.com/espressif/esp-idf/blob/master/components/esp_adc/adc_continuous.c#L199) and [here](https://github.com/espressif/esp-idf/blob/master/components/esp_adc/adc_continuous.c#L215).

While the ISR handlers of the driver **are** IRAM-safe, they are also calling the optional, user-supplied callbacks, which might or might not be IRAM-safe. In my case, they are not and it is way too difficult to make them IRAM-safe.

In other words, besides using the `CONFIG_ADC_CONTINUOUS_ISR_IRAM_SAFE` config [to check whether the user-supplied callbacks are IRAM-safe](https://github.com/espressif/esp-idf/blob/master/components/esp_adc/adc_continuous.c#L600), the code should use the config so as to instantiate the interrupt handling either with the `ESP_INTR_FLAG_IRAM` - when the user-supplied callbacks are IRAM safe - or with `0` - when those are not RAIM-safe.

### What is the actual behavior?

See above.

### Steps to reproduce.

No steps, as this was found by code examination.

**Comment by ivmarkov** [comment]
On a related note, I would appreciate if [this log line](https://github.com/espressif/esp-idf/blob/master/components/esp_adc/adc_continuous.c#L448) is removed. 

While it is only available in verbose logging, its content is misleading if the user is using the driver in non-blocking mode (the support for this (Rust `async`) just [landed](https://github.com/esp-rs/esp-idf-hal/blob/master/src/adc.rs#L907) in `esp-idf-hal`). In non-blocking mode, reading **will always** be with a timeout of 0, and yes - if there is no data, ESP_ERR_TIMEOUT would occur, but that's fine and does not imply the user has to increase the timeout (which she cannot do, as non-blocking implies no waiting and a timeout of 0).

**Comment by ivmarkov** [comment]
Anybody?

@igrr sorry for pinging - can we get somebody to at least look if my assessment is correct?
